### DISCUSSION TITLE
[Bug]: FormProvider -> Portal -> FormProvider mixes form state

### DISCUSSION THREAD

**Discussion by mayteio** [react-hook-form/react-hook-form#6634] [Q&A]
### Version Number

7.15.3

### Codesandbox/Expo snack

https://codesandbox.io/s/holy-rain-nk0oq?file=/src/App.js

### Steps to reproduce

1. Go to https://codesandbox.io/s/holy-rain-nk0oq?file=/src/App.js
2. Click on "open modal" and click submit
3. Main form's validation is triggered inside `useEffect` because `isSubmitted` wrongly set to true

### Expected behaviour

Main FormProvider's state should not be modified by child FormProvider's state. `isSubmitted` should stay `falsy` for main form when portal form is submitted.

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
This is not an issue to the lib, if you submitted the form doesn't matter if it's the main/sub form. the submitted will be updated. there is only a single `useForm` instance unless you have two separate `useForm`. one for main and another for sub form

**Reply by mayteio** [reply]
@bluebill1049 I am using two `useForm` instances, one on line `20` and one on line `71`. Can you please check the codesandbox to verify?

**Reply by bluebill1049** [reply]
miss the second one, have a look at it again.

**Reply by mayteio** [reply]
Thank you. Please let me know if I'm doing something wrong in implementation 👍🏻

**Reply by bluebill1049** [reply]
problem found, i will update both here and in the issue.

**Comment by bluebill1049** [comment]
This is due to nested form: 

```tsx
<form>
  <Modal isOpen={isOpen} onClose={onClose}>
    <ModalOverlay />
    <ModalContent>
      <ModalBody>
        <ModalForm onSuccess={onClose} />
      </ModalBody>
    </ModalContent>
  </Modal>
</form>
```

to 

```tsx
<form>
 ...
</form>
<Modal isOpen={isOpen} onClose={onClose}>
  <ModalOverlay />
  <ModalContent>
    <ModalBody>
      <ModalForm onSuccess={onClose} />
    </ModalBody>
  </ModalContent>
</Modal>
```
          
move the modal out should solve the problem.

if you nested form element with form element, onSubmit will be triggered which leads to `handleSubmit`, this is not related to hook form, instead of native behavior. 

## Work around (not recommend as it breaks accessbility)

you can attach onClick with `handleSubmit` instead to trigger the submission and validation.

**Reply by bluebill1049** [reply]
here is a reproducible CSB: https://codesandbox.io/s/pedantic-tereshkova-xvdsg?file=/src/App.js
<img width="1088" alt="Screen Shot 2021-09-24 at 9 37 21 pm" src="https://user-images.githubusercontent.com/10513364/134668218-eba0b49c-c6b0-4aa6-b0d7-75c0f4eb9655.png">

**Reply by mayteio** [reply]
But then I miss out on handleSubmit validation and the values of the modal form being passed in onSubmit and makes react-hook-form basically useless in the modal.  

Can you please explain how this isn't a bug? The context isn't working as expected. useFormState should be scoped to it's closest parent FormProvider, and especially the control that is passed directly to it as in the main form. 

One would not expect that child FormProvider affects an Ancestor FormProvider - that's not how react context works. Context consumes the closest parent Provider.

**Reply by bluebill1049** [reply]
> makes react-hook-form basically useless in the modal.

refer here... https://github.com/react-hook-form/react-hook-form/discussions/6634#discussioncomment-1379322

please refer to the example above, in the codesandbox it's not even using hook form, how do you expect us to fix it? i even provided a workaround to avoid this problem...

**Reply by mayteio** [reply]
Sorry, there must be some mis-communication. I **am using hook form in both the main form and the modal. See highlighted lines:**

<em>useForm being used in the main component</em>
<img width="737" alt="Screen Shot 2021-09-25 at 9 07 20 am" src="https://user-images.githubusercontent.com/43975092/134748115-2825b01c-68ba-4270-a67a-d3b10a183e98.png">

<em>useForm being used in the modal component</em>
<img width="659" alt="Screen Shot 2021-09-25 at 9 06 04 am" src="https://user-images.githubusercontent.com/43975092/134748137-f4cb1370-8f4b-4b32-a802-a5aa400bd071.png">

It's not a nested `<form>`, because the the `Portal` of the modal moves it outside the component stack. If they were nested, I'd get a native warning about not having nested forms like this:

<img width="1350" alt="Screen Shot 2021-09-25 at 9 09 49 am" src="https://user-images.githubusercontent.com/43975092/134748265-72a06949-9487-4e07-9ae4-d1cf6c02bec8.png">

So two `FormProviders` shouldn't affect each other. [Here's an example of expected behaviour](https://codesandbox.io/s/modest-spence-1t08d?file=/src/App.js) and a [simplified version of the problem](https://codesandbox.io/s/distracted-gates-oh3v4?file=/src/App.js). Shouldn't . Submitting the child form does not affect the parent form's state.

So - **are you saying there is no way to catch the event form submit being bubbled up to the parent form**?

**Reply by mayteio** [reply]
Interestingly, I tried to [write a failing test in Jest](https://github.com/react-hook-form/react-hook-form/pull/6641/files) but it seems it works as expected in JSDOM and passes. This differs from the real environment results (exactly the same code): 

<img width="554" alt="Screen Shot 2021-09-25 at 10 08 59 am" src="https://user-images.githubusercontent.com/43975092/134750938-f788f640-9b5e-43f4-a4bd-329c3f7d50f9.png">

**Comment by mayteio** [comment]
I've fixed it temporarily. By adding `e.stopPropagation()` into the handler:

```tsx
<Portal>
<form
  onSubmit={(e) => {
    e.stopPropagation();
    methods.handleSubmit(() => void 0)(e);
  }}
/>
</Portal>
```

If you'd like to address this, you can add `e.stopPropagation()` [here](https://github.com/react-hook-form/react-hook-form/blob/2a5ca7856c1d31ca86af59769c6b0fb511391073/src/logic/createFormControl.ts#L1021). I've created a PR if you want it.

So just to confirm it **is** a bug with RHF not preventing events propagating up the component chain in a valid HTML structure.

**Reply by bluebill1049** [reply]
did you look at my example above? https://codesandbox.io/s/focused-hooks-r8ly0?file=/src/App.js

it's not even using hook form:
<img width="813" alt="Screen Shot 2021-09-25 at 10 17 42 am" src="https://user-images.githubusercontent.com/10513364/134751277-1b0abfb1-b7d9-41c5-ad9a-7c358caa8e34.png">

This is not an issue related to hook form and we shouldn't add anything in lib to affect this behaviour.

**Reply by mayteio** [reply]
**I need to use RHF in the modal though. For validation, errors, state, etc and getting values `onSubmit`**. I should be able to do this. It's a requirement of our app. You've simply removed the second `useForm` instance from the modal which doesn't solve the problem.

Anyway - thanks for exploring with me. The PR is there that solves this bug. As a workaround (for anyone else who might run into this bug) you can manually call `e.stopPropagation`, then `handleSubmit()(event)` to prevent the portal's form bubbling to the parent.

**Reply by bluebill1049** [reply]
did you look at my example? where is the hook form? someone help plz...
<img width="759" alt="Screen Shot 2021-09-25 at 10 22 03 am" src="https://user-images.githubusercontent.com/10513364/134751439-dc9613e4-8eee-40e3-a482-bf7fb2698ceb.png">

**Reply by bluebill1049** [reply]
I think I have given enough attention and support in here, anyone else can help please help and jump in.

**Reply by mayteio** [reply]
Sorry, I thought you meant in the ModalForm component. I missed that and my apologies. I understand you're saying that it's expected behaviour natively as well. The workaround provided is the incorrect workaround though, using `e.stopPropagation` in `onSubmit` lets you correctly validate a `form` -> `portal` -> `form`.

**Comment by aashishGitHub** [comment]
I had same problem, where I had to show multiple independent forms on the same page to be opened based on some conditions, some as sub-forms. The problem was seen when we submit one form, another form in dom used to get submitted. 
I solved it by using display none to non displayed forms, kept it inside a portal. 
```

<div
  className={clsx('fixed w-full top-0 right-0 transition-all duration-300', {
    'right-[-100%] ': isForm1Closed,
  })}
>
  <div ref={refFormOne} className={isForm1Closed? 'hidden' : 'block'} />
  <Portal container={refFormOne.current}>
    <FormOne />
  </Portal>
</div>

<div
  className={clsx('fixed w-full top-0 right-0 transition-all duration-300', {
    'right-[-100%] ': isForm2Closed,
  })}
>
  <div ref={refFormTwo} className={isForm2Closed? 'hidden' : 'block'} />
  <Portal container={refFormTwo.current}>
    <FormTwo />
  </Portal>
</div>;

```
